### PR TITLE
feat(zk_toolbox): Add prover run

### PR DESCRIPTION
**Pull request by matias-gonz**

### CODE REVIEW

File: zk_toolbox/Cargo.toml [hunk 1]
```diff
@@ -43,7 +43,7 @@ serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 serde_yaml = "0.9"
 sqlx = { version = "0.7.4", features = ["runtime-tokio", "migrate", "postgres"] }
-strum = "0.26.2"
+strum = { workspace = true, features = ["derive"] }
```
Comment (Deniallugo) [inline zk_toolbox/Cargo.toml:46]: it's already a workspace

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    let config_path = chain.path_to_general_config();
+    let secrets_path = chain.path_to_secrets_config();
+
+    let command_str = format!(
+        "cargo run --release --bin zksync_prover_fri_gateway -- --config-path={} --secrets-path={}",
+        config_path.to_str().unwrap(),
+        secrets_path.to_str().unwrap()
+    );
+
+    let pid = run_in_background(&command_str, link_to_prover)?;
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:61]: why in background?

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            ProverComponent::Gateway => run_gateway(&chain, &link_to_prover)?,
+            ProverComponent::WitnessGenerator => run_witness_generator(&chain, &link_to_prover)?,
+            ProverComponent::WitnessVectorGenerator => {
+                run_witness_vector_generator(&chain, &link_to_prover)?
+            }
+            ProverComponent::Prover => run_prover(&chain, &link_to_prover)?,
+            ProverComponent::Compressor => run_compressor(&chain, &link_to_prover)?,
+        }
+    }
+
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:44]: i don't think we have to add the ability to run them simultaneously  in one process

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    Ok(())
+}
+
+fn run_witness_generator(shell: &Shell, chain: &ChainConfig) -> anyhow::Result<()> {
+    logger::info(MSG_RUNNING_WITNESS_GENERATOR);
+    let config_path = chain.path_to_general_config();
+    let secrets_path = chain.path_to_secrets_config();
+
+    let mut cmd = Cmd::new(cmd!(shell, "cargo run --release --bin zksync_witness_generator -- --all_rounds --config-path={config_path} --secrets-path={secrets_path}"));
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:57]: let user specify rounds

File: zk_toolbox/crates/zk_inception/src/commands/prover/run.rs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    cmd.run().context(MSG_RUNNING_WITNESS_VECTOR_GENERATOR_ERR)
+}
+
+fn run_prover(shell: &Shell, chain: &ChainConfig) -> anyhow::Result<()> {
+    logger::info(MSG_RUNNING_PROVER);
+    let config_path = chain.path_to_general_config();
+    let secrets_path = chain.path_to_secrets_config();
+
+    let mut cmd = Cmd::new(
+        cmd!(shell, "cargo run --features gpu --release --bin zksync_prover_fri -- --config-path={config_path} --secrets-path={secrets_path}"),
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/run.rs:77]: i think we have to add `nvidia` as a prerequisite for running prover

File: zk_toolbox/crates/zk_inception/src/commands/prover/init.rs [hunk 6]
```diff
@@ -82,24 +82,15 @@ fn download_setup_key(
         .clone();
     let url = compressor_config.universal_setup_download_url;
 
-    let mut cmd = Cmd::new(cmd!(shell, "wget {url} -P {path}"));
+    let mut cmd = Cmd::new(cmd!(shell, "wget {url} -O {path}"));
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:85]: seems wget is a new requirement

File: etc/env/file_based/general.yaml [hunk 7]
```diff
@@ -315,7 +315,7 @@ house_keeper:
   fri_gpu_prover_archiver_archive_after_secs: 172800
 
 prometheus:
-  listener_port: 3312
+  listener_port: 3116
```
Comment (Deniallugo) [inline etc/env/file_based/general.yaml:318]: we have the same port in api.

File: zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+use strum::{EnumIter, IntoEnumIterator};
+
+use crate::messages::{MSG_ROUND_SELECT_PROMPT, MSG_RUN_COMPONENT_PROMPT};
+
+#[derive(Debug, Clone, Parser, Default)]
+pub struct ProverRunArgs {
+    #[clap(long)]
+    pub component: Option<ProverComponent>,
+    #[clap(flatten)]
+    pub witness_generator_args: WitnessGeneratorArgs,
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/args/run.rs:12]: could we move it to the witness generator? 
I assume we can't run different rounds for gateway

File: zk_toolbox/crates/zk_inception/src/commands/prover/init.rs [hunk 9]
```diff
@@ -135,3 +132,27 @@ fn get_object_store_config(
 
     Ok(object_store)
 }
+
+fn init_bellman_cuda(shell: &Shell) -> anyhow::Result<()> {
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:136]: lets have it as an additional step

File: zk_toolbox/crates/zk_inception/src/commands/prover/init.rs [hunk 10]
```diff
@@ -83,25 +86,19 @@ fn download_setup_key(
         .expect(MSG_PROOF_COMPRESSOR_CONFIG_NOT_FOUND_ERR)
         .clone();
     let url = compressor_config.universal_setup_download_url;
+    let path = std::path::Path::new(path);
+    let parent = path.parent().expect(MSG_SETUP_KEY_PATH_ERROR);
+
+    Cmd::new(cmd!(shell, "wget {url} -P {parent}")).run()?;
+    Cmd::new(cmd!(shell, "mv {parent}/setup_2^24.key {path}")).run()?;
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:93]: Why we can't download it to the necessary directory?

File: zk_toolbox/crates/zk_inception/src/consts.rs [hunk 11]
```diff
@@ -6,3 +6,4 @@ pub const PROVER_MIGRATIONS: &str = "prover/prover_dal/migrations";
 pub const PROVER_STORE_MAX_RETRIES: u16 = 10;
 pub const DEFAULT_CREDENTIALS_FILE: &str = "~/.config/gcloud/application_default_credentials.json";
 pub const DEFAULT_PROOF_STORE_DIR: &str = "artifacts";
+pub const BELLMAN_CUDA_DIR: &str = "era-bellman-cuda";
```
Comment (Deniallugo) [inline zk_toolbox/crates/zk_inception/src/consts.rs:9]: could we add this as a param for ecosystem config?

File: zk_toolbox/crates/zk_inception/src/commands/prover/init.rs (hunk 10 above)
Comment (matias-gonz) [reply zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:93]: We download to the correct directory, the second cmd renames the file to the one the user chose.
I couldn't find a way to download to a specific dir + choose the name of the downloaded file at the same time so I do it in two steps

File: zk_toolbox/crates/zk_inception/src/consts.rs (hunk 11 above)
Comment (matias-gonz) [reply zk_toolbox/crates/zk_inception/src/consts.rs:9]: Definitely 👍

File: zk_toolbox/crates/zk_inception/src/commands/prover/init.rs (hunk 9 above)
Comment (matias-gonz) [reply zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:136]: What do you mean? Another command? Or separate download and installation?

Comment (Deniallugo) [reply zk_toolbox/crates/zk_inception/src/commands/prover/init.rs:140]: I mean could we add a separate command for `download and installation` only

File: zk_toolbox/crates/config/src/ecosystem.rs [hunk 12]
```diff
@@ -25,6 +25,7 @@ struct EcosystemConfigInternal {
     pub name: String,
     pub l1_network: L1Network,
     pub link_to_code: PathBuf,
+    pub bellman_cuda_dir: PathBuf,
```
Comment (Deniallugo) [inline zk_toolbox/crates/config/src/ecosystem.rs:28]: I assume it's suppose to be optional, because we don't need always run and install provers

**Review by Deniallugo — APPROVED** [review]

**Pull request merged**